doc/platform/nlapi/netlink_api_kernel_programming.md

```diff
@@ -0,0 +1,223 @@
+  
+# About this Manual  
+This document provides general information about the SONiC Linux kernel programming performance enhancement using Netlink based APIs.  
+
+# Scope  
+This document describes the high level design of  Linux kernel programming performance enhancement using netlink based API calls replacing system() calls executing iproute2 commands on linux shell. 
+
+# Definition/Abbreviation  
+### Table 1: Abbreviations  
+| **Term** | **Meaning** |  
+|--------------------------|-------------------------------------|  
+| NL | Netlink |  
+| API | Application Program Interface |
+| NLAPI | Netlink API |
+
+
+
+  
+# 1 Requirement Overview  
+## 1.1 Functional Requirements  
+1. Support Netlink API calls to replace all exec calls of iproute2 ip and bridge commands found in SONiC modules. 
+ 
+## 1.2 Configuration and Management Requirements  
+NA
+  
+## 1.3 Scalability Requirements  
+NA
+  
+## 1.4 Warm Boot Requirements  
+NA  
+  
+# 2 Functionality  
+## 2.1 Target Deployment Use Cases  
+Performance improvement in scaled configuration during bulk config or system restart.
+  
+## 2.2 Functional Description  
+This document provides functional design and specifications of kernel programming performance enhancement by replacing system() calls with netlink APIs for kernel interaction.
+ 
+SONiC modules program the kernel constructs like bridge, VLAN, IP-link, VRF, IP Address into kernel using unix iproute-2 utility commands. Within a sonic process, this is being done by forking new process using fork (thru popen) system call and then executing required  iproute2 ip and/or bridge commands in it. This system() call invocation is slow (compared to netlink APIs) and executed synchronously and process executing this waits for its completion, impacting the performance. Kernel programming using this method is primarily used in config managers of SwSS components. Config managers predominantly use the popen call based kernel programming during user configuration or config replay at restart or system bringup. 
+
+Execution or forking a new process inside a module is expensive in terms of CPU load and CPU time.  For instance, during bulk configuration like VLAN range config or during scaled config restart replay, it is observed that invoking fork call based kernel programming uses lot of time to program  the kernel which in tern results in higher completion time within SwSS as well as slower hardware programming.
+
+Iproute2 ip and bridge commands internally use Netlink based messaging to program the kernel constructs like bridge, VLAN, link, VRF etc. 
+
+Objective of this enhancement is to implement the Netlink based Library function calls or APIs instead of using fork based iproute2 command execution. The API calls make use of netlink library to construct netlink messages and configure the kernel as per user request. Library API functions will perform the same functionality as the iproute2 ip and bridge commands invoked by SONiC Modules.
```

Comment:
Are we using native netlink calls or libnl for talking to kernel? we do have libnl wrapper on top of native netlink functions, can you add some details on the internal design to talk to kernel? hope the code can be enhanced easily to add more netlink APIs in the near future?

Reply:
We are using native netlink message calls; Netlink socket and formated messages.
NLAPI netlink kernel interaction implementaion is similar/inline  with how Iproute2 package implements netlink calls. API invoking functions are internally designed and implemented such that they can be expanded easily for any additional attributes.